### ISSUE TITLE
restype = check_error causes problems on Windows environments

### ISSUE THREAD

**Issue by MaxyMoos** [python-xmp-toolkit/python-xmp-toolkit#58]
Hello,

I am currently porting a Python program using libxmp onto Windows 7 and I noticed that I got caught in an infinite loop that doesn't happen on my Linux environment (Ubuntu 14.04):

Using `get_array_item` to walk through a sequence is the culprit, and peculiarly the `check_error` call through the ctypes `restype` variable:
I dumped the value of the `success` parameter while executing the program and it seems like on a Windows environment, when `xmp_get_array_item` returns True, we get success = 1, but when `xmp_get_array_item` returns False, we get... something that is not 1 (various high int values), so the loop of `XMPMeta.count_array_items` goes on indefinitely.

I suspect this is due to the internals of ctypes and ctypes.WinDLL, but the solution I've implemented for now is the following one: replace the `restype = check_error` assignation by `restype = ctypes.c_bool`and call the check_error function on the return value of xmp_get_array_item.

I've only done it for this function in my code, but it might be worth doing this kind of fix in the whole lib in order to enhance Windows compatibility.
Anyway it's up to you guys to decide if you think that's worth it !

Workaround code:

```
def get_array_item(xmp, schema, name, index):
    EXEMPI.xmp_get_array_item.restype = ctypes.c_bool
    EXEMPI.xmp_get_array_item.argtypes = [ctypes.c_void_p,
                                          ctypes.c_char_p,
                                          ctypes.c_char_p,
                                          ctypes.c_int32,
                                          ctypes.c_void_p,
                                          ctypes.POINTER(ctypes.c_uint32)]
    _item = _string_new()
    property_bits = ctypes.c_uint32(0)

    retVal = EXEMPI.xmp_get_array_item(xmp,
                              schema.encode('utf-8'),
                              name.encode('utf-8'),
                              ctypes.c_int32(index),
                              _item,
                              ctypes.byref(property_bits))

    check_error(retVal)
    item = string_cstr(_item)
    _string_free(_item)

    return item, property_bits.value
```

**Comment by MaxyMoos** [comment]
I've just noticed that a similar issue was described in issue #15, about old versions of exempi, but I'm using the latest exempi version available on GitHub (pulled from the master branch) on my Windows environment, so this current issue may still be valid.

**Comment by quintusdias** [comment]
Might have to ask you for your instructions on how to build exempi on windows so I can test it.  I've tried building it a few times, always gave up.

**Comment by MaxyMoos** [comment]
You're right, that was most definitely not a straightforward process.

I've set up a repo where I put all my Windows-compatible exempi directory : https://github.com/MaxyMoos/exempi-Win7-VS2013/

Here are the steps that lead to a successful compilation on my machine:
Everything was built with Visual Studio Express 2013 on Windows 7 (64-bit), so there might be some instructions that apply only to this specific version:

**1-** In _/build/shared/CMakeUtils.bat_, change

`set GeneratorVersion = Visual Studio 11` **to** `set GeneratorVersion = Visual Studio 12` (line 25)

and (lines 62 to 64):
```
if /I "%1"=="2011" (
echo "Generator VS 2011 specified"
set GeneratorVersion=Visual Studio 11
```
to:
```
if /I "%1"=="2013" (
echo "Generator VS 2013 specified"
set GeneratorVersion=Visual Studio 13
```

**2-** Add the zlib (1.2.8) headers and expat (2.1.1) lib folder in _/third-party/zlib/_ and _/third-party/expat/_

**3-** In /build, execute the GenerateXMPToolkitSDK_win.bat tool and use option 2 (_Generate XMPSDKToolkit Dynamic Win32_). The cmake build should now work and Visual Studio solutions should have been generated for XMPCore & XMPFiles.

**4-** Beginning the fun part: there are some modifications to be done in order to compile the XMPFiles solution:

In _XMPFiles/source/FileHandlers/JPEG_Handler.cpp_ :
Add `#include <algorithm>` after all the other include instructions (else there's a conflict for the definitions of std::min and std::max)

And we need to add a few files that are missing from the project...
In the XMPFiles project, open the Solution Explorer and under XMPFiles / Header Files / Internal Headers:
* Add under File Handlers: GIF_Handler.hpp
* Add under Format Support: GIF_Support.hpp

And under XMPFiles / Source Files :
* Add under File Handlers: GIF_Handler.cpp
* Add under Format Support: GIF_Support.cpp

**5-** Compile both XMPFiles and XMPCore solutions

**6-** Create a new project _libexempi_ in VS and add the following files from the _/exempi_ folder to it:
- exempi.cpp
- xmp.h
- xmp++.hpp
- xmpconsts.h
- xmperrors.h

In the project properties, add to the Include folders:
* The pthread headers directory
* The XMPCore and XMPFiles source directories
* The exempi root directory
* The /public folder of the exempi root dir

Add to the linker libraries folders:
* The pthread lib directory
* The /public/libraries/windows/Release directory, where your XMPCore and XMPFiles DLLs and .lib must have been generated

Add to the linker's _Additional dependencies_ :
* XMPFiles.lib
* XMPCore.lib
* pthreadVSE2.lib

Finally, in /public/XMP_Environment.h, add the following at the top of the file:
`#define WIN_ENV 1`

And last but not least: add `DllExport` around all functions in xmp.h and exempi.cpp.
I strongly recommend you get those two files from my repo as it:
1. Is not very interesting to copy/pasting DllExport statements... and I've done it already so take advantage
2. I encountered a problem with the CompareDateTime function call (linker issue that I didn't manage to solve), so to work around it, I duplicated all the necessary code in exempi.cpp

Don't forget to set the project configuration to generate a DLL and not a .exe or static lib and normally, you're good to go.

Hope that'll help !

**Comment by zougloub** [comment]
A problem with check_error was corrected but Windows support isn't available (see #83) and the issue is quite old, so I'm closing this ; note for those that got exempi to compile on Windows that sharing / upstreaming the recipe could be welcome by Exempi upstream. Please free to reopen if the issue is re-demonstrated.
